cprogs/with-lock-ex.1
.TH WITH-LOCK-EX "1" "July 2003" "Debian" "Chiark-utils-bin"
.SH NAME
with-lock-ex \- file locker
.SH SYNOPSIS
.B with-lock-ex
.BR \-w \||\| \-q \||\| \-f
.I lockfile command 
.IR args \ \|.\|.\|.
.br
.SH DESCRIPTION
with-lock-ex will open and lock the lockfile for writing and then feed
the remainder of its arguments to
.BR exec (2);
when that process terminates the fd will be closed and the file
unlocked automatically by the kernel.
.PP
If the file does not exist it is created, with permissions
.B rw
for each user class for which the umask has
.BR w .
.SH OPTIONS
.TP
.B \-w
Wait for the lock to be available.
.TP
.B \-f
Fail (printing a message to stderr and exiting 255) if the lock cannot
be acquired immediately because another process has it.
.TP
.B \-q
Silently do nothing (ie, exit 0 instead of executing the specified
process) if the lock cannot be acquired immediately because another
process has it.
.SH STALE LOCKS
The locking protocol used does not suffer from stale locks.  If the
lock cannot be acquired, one or more running processes must currently
hold the lock; if the lock needs to be freed those processes should be
killed.
.PP
Under no circumstances should `stale lock cleaner' cron jobs, or the
like, be instituted.  In systems where a great many locks may exist,
old lockfiles may be removed from cron but only if each lock is
acquired before the lockfile is removed, for example with
.IP
.B with-lock-ex -q
.I lockfile
.B rm
.I lockfile
.SH DEADLOCKS
There is no deadlock detection.  In a system with several locks, a
lock hierarchy should be established, such that for every pair of
locks
.I A
and
.I B
which a process might lock simultaneously, either
.IR A > B
or
.IR B > A
where the relation > is transitive and noncyclic.
.PP
Then, for any two locks
.I X
and
.I Y
with
.IR X > Y
it is forbidden to acquire
.I X
while holding
.IR Y .
Instead, acquire
.I X
first, or release
.I Y
before (re)acquiring
.I X
and
.I Y
in that order.
.PP
(There are more complicated ways of avoiding deadlocks, but a lock
hierarchy is simple to understand and implement.  If it does not meet
your needs, consult the literature.)
.SH LOCKING PROTOCOL
The locking protocol used by
.B with-lock-ex
is as follows:
.PP
The lock is held by a process (or group of processes) which holds an
fcntl exclusive lock on the first byte of the plain file which has the
specified name.  A holder of the lock (and only a holder of the lock)
may delete the file or change the inode to which the name refers, and
as soon as it does so it ceases to hold the lock.
.PP
Any process may create the file if it does not exist.  There is no
need for the file to contain any actual data.  Indeed, actually using
the file for data storage is strongly disrecommended, as this will
foreclose most strategies for reliable update.  Use a separate
lockfile instead.
.PP
Ability to obtain the lock corresponds to write permission on the file
(and of course permission to create the  file, if it does not already
exist).  However, processes with only read permission on the file can
prevent the lock being acquired at all; therefore lockfiles should not
usually be world-readable.
.PP
When a (group of) processes wishes to acquire the lock, it should open
the file
(with
.BR O_CREAT )
and lock it with
.BR fcntl (2)
.BR F_RWLCK ,
operation
.B F_SETLK
or
.BR F_SETLKW .
If this succeeds it should fstat the file descriptor it has, and the
file by its path.  If the device and inode match then the lock has
been acquired and remains acquired until that group of processes
changes which file the name refers to, deletes the file, or releases
the fcntl lock.  If they do not then another process acquired the lock
and deleted the file in the meantime; you must now close your
filedescriptor and start again.
.B with-lock-ex
follows this specification.
.PP
Note that
.BR flock (2)
is a different kind of lock to
.BR fcntl (2).
.B with-lock-ex
uses
.BR fcntl .
.SH AUTHOR
This Manual page was written by Matthew Vernon <matthew@debian.org>
and enhanced by Ian Jackson <ian@chiark.greenend.org.uk>, in 2003, but
may be used by anyone.
.SH COPYRIGHT
with-lock-ex was written by Ian Jackson <ian@chiark.greenend.org.uk>
in 1993, 1994, 1995, 1996, 1998, 1999. He has placed it in the public
domain. 
.SH "SEE ALSO"
.BR fcntl (2),
.BR flock (2),
.BR chmod (2)


debian/changelog
chiark-utils (8.0.0-0deepin1) unstable; urgency=medium

  * feat: add sw64 support 

 -- hudeng <hudeng@deepin.org>  Tue, 21 Oct 2025 17:32:41 +0800

chiark-utils (8.0.0) unstable; urgency=medium

  * Rewrite the packaging to use `debhelper` and `dh` in as standard a way
    as possible.  As a consequence, support root-less builds.
    No functional changes to the installed packages.
    Closes: #1089303.  [Niels Thykier]

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 12 Jan 2025 20:19:58 +0000

chiark-utils (7.0.3) unstable; urgency=medium

  * Team upload.
  * git-cache-proxy: Handle extra-parameters in git protocol (Closes: #1040476).
    Thanks to Anthony PERARD for the patch.
  * git-branchmove: Always pass --no-verify to 'git push' (Closes: #1080082).

 -- Sean Whitton <spwhitton@spwhitton.name>  Thu, 26 Sep 2024 17:07:41 +0100

chiark-utils (7.0.2) unstable; urgency=medium

  * debian/rules: Pass -D_TIME_BITS=64 too.  Closes: #1078190.
  * Change file size type to unsigned long long.  Closes: #1075821.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 15 Aug 2024 22:28:46 +0100

chiark-utils (7.0.1) unstable; urgency=medium

  * backup/snaprsync: add -AX to the rsync rune
  * Enable LFS and 64-bit time_t support.  Closes: #1067157.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 04 Jul 2024 12:53:58 +0100

chiark-utils (7.0.0) unstable; urgency=medium

  New facility:
  * prefork-interp, Proc::Prefork::Interp: new facility

  Improvements:
  * expire-iso8601: Replace with new implementation.  Closes:#862897.

  Bugfixes:
  * acctdump: Fix manual page section.
  * git-cache-proxy: fix for new git.  Closes: #1006695.  [Anthony Perard]
  * git-branchmove: install manpage..  Closes: #946364.  [Sean Whitton]
  * hexterm: Update #! and Suggest tcl8.6.  Closes: #905474.

  Build and packaging improvements:
  * Fix cross build.  Closes: #951242.  [Helmut Grohne]
  * copyright: Be clear that SPDX indications are licence grants
  * Add lintian override for python dependency
  * Add lintian override for chiark-really
  * fix priority "extra" to "optional"
  * Fix build-dependency on debhelper.
  * Set source format to 3.0 (native)
  * Add Vcs-* fields.
  * Update Standards-Version
  * timespeccmp.h: Move use to just cgi-fcgi-interp

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 22 Aug 2022 00:06:57 +0100

chiark-utils (6.1.2+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * Bump debhelper compat to 13 (Closes: #965453).

 -- Sean Whitton <spwhitton@spwhitton.name>  Tue, 21 Jul 2020 08:17:49 -0700

chiark-utils (6.1.2) unstable; urgency=medium

  * git-cache-proxy: Recheck gc timestamp after taking lock
  * git-cache-proxy: Tolerate failure to write gc stamp
  * git-cache-proxy: Do not update timestamp due to housekeeping gc

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Wed, 12 Feb 2020 16:10:02 +0000

chiark-utils (6.1.2~citrix2~) unstable; urgency=medium

  * git-cache-proxy: Move $SIG{ALRM} setting to global
  * git-cache-proxy: Abandon fetch attempt if client disconnects
  * git-cache-proxy: Use open-coded fork/exec for git gc

 -- Ian Jackson <ian.jackson@citrix.com>  Tue, 11 Feb 2020 18:34:58 +0000

chiark-utils (6.1.1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module
  * git-cache-proxy: Periodically run `git gc --quiet'
  * git-cache-proxy: Document some options in the head comment

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 11 Feb 2020 16:37:31 +0000

chiark-utils (6.0.5~citrix1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module, not "os.errno"
    (which is not in modern python3 apparently).

 -- Ian Jackson <ian.jackson@citrix.com>  Fri, 20 Dec 2019 17:50:02 +0000

chiark-utils (6.1.0~iwj1) unstable; urgency=medium

  * really: Writeability of rc.local suffices, too.
    (Non-sysvinit compatibility.)

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 12 Aug 2019 10:59:02 +0100

chiark-utils (6.1.0) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * git-branchmove: rewrite in perl (Closes: #914398, #914399)
    - Add dependencies on libgit-wrapper-perl, libtry-tiny-perl to
      bin:chiark-scripts.
  * git-branchmove: new --detach feature.

 -- Sean Whitton <spwhitton@spwhitton.name>  Sat, 07 Dec 2019 14:10:26 -0700

chiark-utils (6.0.4) unstable; urgency=medium

  * sync-accounts: Fix perl syntax error.  Closes:#865985.
  * changelog: Document bug number for bugfix in 6.0.4~citrix1.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 12 Apr 2019 11:48:00 +0100

chiark-utils (6.0.4~citrix1) unstable; urgency=medium

  * fishdescriptor: cast __errno_location correctly.  Closes:#926858.

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 08 Apr 2019 17:03:47 +0100

chiark-utils (6.0.3) unstable; urgency=medium

  * Upload to Debian unstable.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 20 Nov 2018 18:25:53 +0000

chiark-utils (6.0.3~citrix1) unstable; urgency=medium

  * git-branchmove: Avoid transporting tags about
  * fishdescriptor: Cope if donor is in a chroot without /proc
  * fishdescriptor: Cope if only donor has /run/user/UID

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 19 Nov 2018 16:11:55 +0000

chiark-utils (6.0.2) unstable; urgency=medium

  * Rebuild with arch:all packages included in the upload.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:35:08 +0100

chiark-utils (6.0.1) unstable; urgency=low

  fishdescriptor:
  * Fix numerous bugs, affecting the `exec' feature with nontrivial fd
    specifications.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:31:24 +0100

chiark-utils (6.0.0) unstable; urgency=medium

  xbatmon-simple:
  * Minor improvements to the .txt doc.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 20 Apr 2018 16:56:04 +0100


cprogs/with-lock-ex.c
/*
 * File locker
 *
 * Usage:
 *  with-lock-ex -<mode> [-t <secs>] <lockfile> <command> <args>...
 *  with-lock-ex -l      <lockfile>
 *
 * modes are
 *  w    wait for the lock
 *  f    fail if the lock cannot be acquired
 *  q    silently do nothing if the lock cannot be acquired
 *  l    show who is waiting (print "none" or "read <pid>"
 *         or "write <pid>"; lockfile opened for reading;
 *         no command may be specified)
 *
 * If -t is specified, then with-lock-ex will wait for up to <secs>
 * seconds to acquire the lock, and then fail or silently do nothing
 * (depending on whether -f or -q is specified). You cannot specify
 * a timeout for modes l or w.
 *
 * with-lock-ex will open and lock the lockfile for writing and
 * then feed the remainder of its arguments to exec(2); when
 * that process terminates the fd will be closed and the file
 * unlocked automatically by the kernel.
 *
 * If invoked as with-lock, behaves like with-lock-ex -f (for backward
 * compatibility with an earlier version).
 *
 * This file written by me, Ian Jackson, in 1993, 1994, 1995, 1996,
 * 1998, 1999, 2016.
 *
 * Copyright 1993-2016 Ian Jackson in some jurisdictions
 * Copyright 2017      Ian Jackson in all jurisdictions
 * Copyright 2017      Genome Research Ltd
 *
 * (MIT licence:)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
 * SOFTWARE IN THE PUBLIC INTEREST, INC. BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include <limits.h>
#include <signal.h>
#include <sys/time.h>

static const char *cmd;

static void fail(const char *why) __attribute__((noreturn));

static void fail(const char *why) {
  fprintf(stderr,"with-lock-ex %s: %s: %s\n",cmd,why,strerror(errno));
  exit(255);
}

static void badusage(void) __attribute__((noreturn));

static void badusage(void) {
    fputs("usage: with-lock-ex -w|-q|-f [-t <secs>] <lockfile> <command> <args>...\n"
	  "       with-lock-ex -l       <lockfile>\n"
	  "       with-lock-ex          <lockfile> <command> <args>...\n",
	  stderr);
    exit(255);
}

static int mode;

/* This signal handler uses unsafe functions, so MUST NOT be callable
 * during an unsafe function, as that is Undefined Behaviour
 */
static void alrm_handler(int signum) {
  if (mode=='q') {
    exit(0);
  } else {
    fprintf(stderr,
	    "with-lock-ex %s: timer expired while trying to acquire lock\n",
	    cmd);
    exit(255);
  }
}

int main(int argc, char **argv) {
  int fd, um, c, r;
  struct stat stab, fstab;
  long cloexec, secs=0;
  struct flock fl;
  char *endptr;
  sigset_t sigs, oldsigs;
  struct sigaction siga;
  struct itimerval itv;

  mode= 'x';
  while ((c= getopt(argc,argv,"+wfqlt:")) != -1) {
    switch(c) {
    case 'l':
    case 'w':
    case 'f':
    case 'q':
      if (mode != 'x') badusage();
      mode= c;
      break;
    case 't':
      errno = 0;
      secs = strtol(optarg, &endptr, 0);
      if (*endptr || endptr==optarg || errno==ERANGE)
	fail("parsing timeout value");
      if (secs < 0) {
	fprintf(stderr,"timeout value must be >=0\n");
	exit(255);
      }
      break;
    default:
      badusage();
    }
  }

  if (secs && (mode=='l' || mode=='w')) {
    fputs("-t only allowed with -q or -f.\n", stderr);
    exit(255);
  }

  argv += optind-1; argc -= optind-1;
  if (argc < 2) badusage();

  if (secs) {
    if (sigemptyset(&sigs)) fail("Initialising signal set");
    if (sigaddset(&sigs,SIGALRM)) fail("Adding SIGALRM to signal set");
    if (sigprocmask(SIG_BLOCK,&sigs,&oldsigs)) fail("Blocking SIGALRM");
    memset(&siga,0,sizeof(siga));
    siga.sa_handler=alrm_handler;
    if (sigaction(SIGALRM,&siga,NULL)) fail("Installing SIGALRM handler");
    memset(&itv,0,sizeof(itv));
    itv.it_value.tv_sec=secs;
    if (setitimer(ITIMER_REAL,&itv,NULL)) fail("Setting timer");
  }

  cmd= argv[2];
  um= umask(0777); if (um==-1) fail("find umask");
  if (umask(um)==-1) fail("reset umask");

  for (;;) {

    int openmode = mode=='l' ? O_RDONLY : O_RDWR|O_CREAT;

    fd= open(argv[1],openmode,0666&~(um|((um&0222)<<1)));
    if (fd<0) fail(argv[1]);
  
    for (;;) {
      fl.l_type= F_WRLCK;
      fl.l_whence= SEEK_SET;
      fl.l_start= 0;
      fl.l_len= mode=='l' ? 0 : 1;
      if (secs) sigprocmask(SIG_UNBLOCK,&sigs,NULL);
      r = fcntl(fd,
		mode=='l' ? F_GETLK :
		mode=='w' || secs > 0 ? F_SETLKW :
		F_SETLK,
		&fl);
      if (secs) sigprocmask(SIG_BLOCK,&sigs,NULL);
      if (!r) {
	break;
      }
      if (mode=='q' &&
	  (errno == EAGAIN || errno == EWOULDBLOCK || errno == EBUSY))
	exit(0);
      if (errno != EINTR) fail("could not acquire lock");
    }
    if (mode=='l') {
      if (fl.l_pid) {
	printf("%s %lu\n",
	       fl.l_type == F_WRLCK ? "write" :
	       fl.l_type == F_RDLCK ? "read" : "unknown",
	       (unsigned long)fl.l_pid);
      } else {
	printf("none\n");
      }
      if (ferror(stdout)) fail("print to stdout\n");
      exit(0);
    }

    if (fstat(fd, &fstab)) fail("could not fstat lock fd");
    if (stat(argv[1], &stab)) {
      if (errno != ENOENT) fail("could not stat lockfile");
    } else {
      if (stab.st_dev == fstab.st_dev &&
	  stab.st_ino == fstab.st_ino) break;
    }
    close(fd);
  }

  if (secs) {
    itv.it_value.tv_sec=0;
    if (setitimer(ITIMER_REAL,&itv,NULL)) fail("Clearing timer");
    sigprocmask(SIG_SETMASK,&oldsigs,NULL);
    siga.sa_handler=SIG_DFL;
    sigaction(SIGALRM,&siga,NULL);
  }

  cloexec= fcntl(fd, F_GETFD); if (cloexec==-1) fail("fcntl F_GETFD");
  cloexec &= ~1;
  if (fcntl(fd, F_SETFD, cloexec)==-1) fail("fcntl F_SETFD");

  execvp(cmd,argv+2);
  fail("unable to execute command");
}
